Source/Core/Common/Arm64Emitter.cpp

```diff
@@ -1992,97 +1993,114 @@ void ARM64XEmitter::ADRP(ARM64Reg Rd, s32 imm)
   EncodeAddressInst(1, Rd, imm >> 12);
 }
 
-// Wrapper around MOVZ+MOVK (and later MOVN)
-void ARM64XEmitter::MOVI2R(ARM64Reg Rd, u64 imm, bool optimize)
+template <typename T>
+void ARM64XEmitter::MOVI2RImpl(ARM64Reg Rd, T imm)
 {
-  unsigned int parts = Is64Bit(Rd) ? 4 : 2;
-  BitSet32 upload_part(0);
-
-  // Always start with a movz! Kills the dependency on the register.
-  bool use_movz = true;
-
-  if (!imm)
+  enum class Approach
   {
-    // Zero immediate, just clear the register. EOR is pointless when we have MOVZ, which looks
-    // clearer in disasm too.
-    MOVZ(Rd, 0, ShiftAmount::Shift0);
-    return;
-  }
+    MOVZ,
+    MOVN,
+    ADR,
+    ADRP,
+  };
 
-  if ((Is64Bit(Rd) && imm == std::numeric_limits<u64>::max()) ||
-      (!Is64Bit(Rd) && imm == std::numeric_limits<u32>::max()))
+  struct Part
   {
-    // Max unsigned value (or if signed, -1)
-    // Set to ~ZR
-    ARM64Reg ZR = Is64Bit(Rd) ? SP : WSP;
-    ORN(Rd, ZR, ZR, ArithOption(ZR, ST_LSL, 0));
-    return;
-  }
+    Part() = default;
+    Part(u16 imm_, ShiftAmount shift_) : imm(imm_), shift(shift_) {}
 
-  // TODO: Make some more systemic use of MOVN, but this will take care of most cases.
-  // Small negative integer. Use MOVN
-  if (!Is64Bit(Rd) && (imm | 0xFFFF0000) == imm)
-  {
-    MOVN(Rd, ~imm, ShiftAmount::Shift0);
-    return;
-  }
+    u16 imm;
+    ShiftAmount shift;
+  };
 
-  // XXX: Use MOVN when possible.
-  // XXX: Optimize more
-  // XXX: Support rotating immediates to save instructions
-  if (optimize)
-  {
-    for (unsigned int i = 0; i < parts; ++i)
-    {
-      if ((imm >> (i * 16)) & 0xFFFF)
-        upload_part[i] = 1;
-    }
-  }
+  constexpr size_t max_parts = sizeof(T) / 2;
 
-  u64 aligned_pc = (u64)GetCodePtr() & ~0xFFF;
-  s64 aligned_offset = (s64)imm - (s64)aligned_pc;
-  // The offset for ADR/ADRP is an s32, so make sure it can be represented in that
-  if (upload_part.Count() > 1 && std::abs(aligned_offset) < 0x7FFFFFFFLL)
-  {
-    // Immediate we are loading is within 4GB of our aligned range
-    // Most likely a address that we can load in one or two instructions
-    if (!(std::abs(aligned_offset) & 0xFFF))
+  Approach best_approach;
+  std::vector<Part> best_parts(max_parts + 1);
```

Comment:
Consider a comment here as this may not be obvious. Maybe (? rewording required): 
```suggestion
  // Have best_parts.size() initially be big to ensure first try_base call replaces it.
  std::vector<Part> best_parts(max_parts + 1);
```